When compiling C++, output a diagnostic per C++ diagnostic

bricknerb opened the pull request:
Part of #5176.

bricknerb committed:
When compiling C++, output a diagnostic per C++ diagnostic
Part of #5176.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,71 @@ static auto GenerateCppIncludesHeaderCode(
+    llvm::SmallString<100> message;
+    info.FormatDiagnostic(message);
+
+    std::string diagnostics_str;
+    llvm::raw_string_ostream diagnostics_stream(diagnostics_str);
```
Had you considered using RawStringOstream here?

bricknerb replied:
Done. Thanks.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,71 @@ static auto GenerateCppIncludesHeaderCode(
 
+namespace {
+
+class CarbonClangDiagnosticConsumer : public clang::DiagnosticConsumer {
```
Can you add comments to this class and its members?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,71 @@ static auto GenerateCppIncludesHeaderCode(
+      }
+      case clang::DiagnosticsEngine::Warning: {
+        CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning,
+                          "Warning in imported C++ code:\n{0}", std::string);
```
Stylistically, I think this should probably look something more like "in C++ import". It would be more similar to the `InImport` diagnostic if phrased that way.

Regarding specifics, the capitalization and `:` don't match [Carbon's diagnostic style guidance](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide).

Note too, in this message:

```
// CHECK:STDERR: import_cpp_file_with_one_warning.carbon:[[@LINE+9]]:1: warning: Warning in imported C++ code:
// CHECK:STDERR: In file included from import_cpp_file_with_one_warning.carbon.generated.cpp_imports.h:1:
// CHECK:STDERR: ./one_warning.h:2:2: warning: "warning1"
```

Both Carbon and Clang are adding "warning:" text, so adding it a third time in the message isn't needed.

bricknerb replied:
Done.

jonmeow replied:
It looks like you've significantly changed the diagnostic structure. I want to note, I think the prior setup here was pretty good; I was only suggesting "Warning in imported C++ code:\n{0}" -> "in C++ import\n{0}". The location output can be improved, but that's more complex.

To help explain, can you please add a test with an error in an indirect include, e.g.:

```
// --- indirect_error.h

#include "one_error.h"

// --- fail_import_cpp_file_with_indirect_error.carbon

library "[[@TEST_NAME]]";

import Cpp library "indirect_error.h"
```

With your current diagnostic structure (with `InCppImport`), I think the most important things to fix are:

1. The repeated "In file included from" case leads to different indentation, which may be hard to read.
  - Suggestion: Put a newline before the Clang message so that its contents are all at the same indent.
2. The Carbon import line is reported as a location twice.
  - Suggestion: It should only be used as a diagnostic location once.
3. We use `AnnotationScope` when diagnostics don't know whether they need a note.
  - Suggestion: Assuming the scope only covers the diagnostics you're adding, they can directly add any notes they want, without indirection.

---

If it helps to think about where we want to end up, I really think the `InImport` example is the best guideline. Here's something we do right now:

```
fail_todo_merge_define_extern.carbon:3:1: note: in import
other_extern.carbon:4:1: error: duplicate name `C` being declared in the same scope [NameDeclDuplicate]
extern class C;
^~~~~~~~~~~~~~~
```

Correspondingly, C++ diagnostics might look like:

```
fail_import_cpp_file_with_multiple_errors.carbon:3:1: note: in import
In file included from fail_import_cpp_file_with_multiple_errors.carbon.generated.cpp_imports.h:1:
./multiple_errors.h:3:2: error: "error2"
    3 | #error "error2"
      |  ^
```

But that kind of format is something that we will only get by improving the C++ location information, and having the diagnostic emitter add import locations in [the same way it does for `InImport`](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/check/sem_ir_loc_diagnostic_emitter.cpp#L51-L54).

bricknerb replied:
Thanks for clarifying!

I've added indirect tests.

For the fixes you suggested:
1. Adding a "\n" before the message seems to make the line end with a space and generate "{{}}" escaping before the newline in the tests. I think adding some string before the newline is better so I added "C++:". WDYT?
2. Since we don't have a better location (yet) I think it's either removing `InCppImport` or remove the location from the warning/errors. I've removed it from the warning/errors for now.
3. Oh, I thought it's useful when we want to make sure all diagnostics in this scope would have this note, so we don't have to remember to replicate adding that note for each diagnostic (and add code duplication). Moved that note to the warning/error emitting diagnostics.

BTW, the link you put seems to have been broken by #5188.
I assume you referred to [this code](https://github.com/carbon-language/carbon-lang/blob/0a3efb76edac15771f7035cd01aa922152cb4d82/toolchain/check/diagnostic_emitter.cpp#L50-L53). I suggest using permalinks to make it more stable for the review and future readers of the PR comments.

jonmeow replied:
On (1) and (2), seems reasonable for now. I've suggested TODOs to clarify what should change in the future.

> 3. Oh, I thought it's useful when we want to make sure all diagnostics in this scope would have this note, so we don't have to remember to replicate adding that note for each diagnostic (and add code duplication). Moved that note to the warning/error emitting diagnostics.

If you can write a test demonstrating the benefit of having this be scoped, I could review that. If there were a specific plan that would take advantage of this, then maybe. I don't think either applies here, so simpler code is better. This code in particular should go away once the location is improved, so there's an added benefit of keeping it next to the location.

bricknerb replied:
SG

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,71 @@ static auto GenerateCppIncludesHeaderCode(
+                        const clang::Diagnostic& info) override {
+    DiagnosticConsumer::HandleDiagnostic(diag_level, info);
+
+    llvm::SmallString<100> message;
```
Had you considered 256 here, rather than 100? Note that there are trade-offs here, like memory alignment, where a power-of-2 choice might do better. Also, diagnostic messages seem likely to frequently be over 100 chars; some might fit into 256, and comparing with other code I see at least some doing 256 for diagnostic messages. (example: https://github.com/llvm/llvm-project/blob/main/clang/lib/Frontend/SerializedDiagnosticPrinter.cpp#L576)

bricknerb replied:
Changed to 256.
I saw 100 is pretty popular as well but 256 makes sense.
I wonder if there's a good way to measure this to find a good value for each `SmallString`.

jonmeow replied:
In general we'd probably only notice if it's a bottleneck. Here, I think it'd need to be seen how many diagnostics are typically produced that are short enough for stack allocation.

bricknerb replied:
Makes sense.
That discussion probably belongs outside of this PR but I wonder if we could have some option to run a benchmark that will enable some `SmallString` debugging and can provide recommendations on cases we may want to change their value.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,71 @@ static auto GenerateCppIncludesHeaderCode(
+      case clang::DiagnosticsEngine::Remark: {
+        context_.TODO(
+            loc_, llvm::formatv(
+                      "Unsupported: C++ diagnostic level for diagnostic {0}",
```
```suggestion
                      "Unsupported: C++ diagnostic level for diagnostic\n{0}",
```
The diagnostic message may be pretty long, so perhaps give it a linebreak as a separator?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,71 @@ static auto GenerateCppIncludesHeaderCode(
+  CarbonClangDiagnosticConsumer(Context& context, SemIRLoc loc)
+      : context_(context), loc_(loc) {}
+
+  void HandleDiagnostic(clang::DiagnosticsEngine::Level diag_level,
```
We're trying to use trailing returns; clang-tidy won't catch void, but can you please switch this?

bricknerb replied:
Sure, done.
My brain still needs to adjust to this style to make sure my eye hurt when it's wrong :)

jonmeow reviewed (commented):
Thanks, I think this is the right direction.

bricknerb committed:
Use `RawStringOstream`

bricknerb committed:
Add comments

bricknerb committed:
Unify Clang warnings and errors and add `InCppImport` context

bricknerb committed:
Change diagnostic message `SmallString` size to 256.

bricknerb committed:
Add newline before unsupported diagnostic message and use trailing return

bricknerb committed:
Add indirect C++ warning test and indirect C++ error test

bricknerb committed:
Add "C++\n" before the Clang error/warning message

bricknerb committed:
Remove `DiagnosticAnnotationScope` and make the `InCppImport` explicit.
Remove the location that was used twice from the warning/error and leave it only on the `Note`.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -72,38 +75,46 @@ class CarbonClangDiagnosticConsumer : public clang::DiagnosticConsumer {
+                          std::string);
+        CARBON_DIAGNOSTIC(CppInteropParseError, Error, "C++:\n{0}",
+                          std::string);
+        CARBON_DIAGNOSTIC(InCppImport, Note, "in `Cpp` import");
```
```suggestion
        // TODO: This should be part of the location, instead of added as a note here.
        CARBON_DIAGNOSTIC(InCppImport, Note, "in `Cpp` import");
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,82 @@ static auto GenerateCppIncludesHeaderCode(
+// Used to convert Clang diagnostics to Carbon diagnostics.
+class CarbonClangDiagnosticConsumer : public clang::DiagnosticConsumer {
+ public:
+  // Creates an instance with the location that triggers calling Clang.
+  CarbonClangDiagnosticConsumer(Context& context, SemIRLoc loc)
```
```suggestion
  // Creates an instance with the location that triggers calling Clang. `context` is required.
  explicit CarbonClangDiagnosticConsumer(Context* context, SemIRLoc loc)
```

Style, per https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md:

- explicit constructors: "Always mark constructors explicit unless there's a specific reason to support implicit or {} initialization."
- Pointer: "If it is captured and must outlive the call expression itself, use a pointer and document that it must not be null"

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,82 @@ static auto GenerateCppIncludesHeaderCode(
+
+ private:
+  // The type-checking context in which we're running Clang.
+  Context& context_;
```
```suggestion
  Context* context_;
```

Style: "When storing an object's address as a non-owned member, prefer storing a pointer."

bricknerb replied:
Thanks! I actually prefer pointers in these cases but saw that it's very popular to keep references in Carbon codebase.
Done and changed the constructor parameter to take a pointer as well, similar to the example in the style guide.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -44,6 +44,82 @@ static auto GenerateCppIncludesHeaderCode(
+      case clang::DiagnosticsEngine::Warning:
+      case clang::DiagnosticsEngine::Error:
+      case clang::DiagnosticsEngine::Fatal: {
+        CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning, "C++:\n{0}",
```
```suggestion
        // TODO: Adjust diagnostics to drop the Carbon file here, and then remove the "C++:\n" prefix.
        CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning, "C++:\n{0}",
```

bricknerb replied:
Done.

jonmeow reviewed (approved):
Approving, assuming you address remaining comments. I think I'd overlooked the constructor before, but should be straightforward to fix.

Thanks!

bricknerb committed:
Change `CarbonClangDiagnosticConsumer::context_` from reference to pointer

bricknerb committed:
Add a comment that `context` parameter must not be null

bricknerb committed:
Make `CarbonClangDiagnosticConsumer` constructor explicit

bricknerb committed:
Add TODOs to improve diagnostics

bricknerb committed:
Merge branch 'main' into diag
